File: src/my_robot_description/scripts/robot_controller.py
#!/usr/bin/env python3
import rclpy
from rclpy.node import Node
from geometry_msgs.msg import Twist

class RobotController(Node):
    def __init__(self):
        super().__init__('robot_controller')
        # This publisher talks to the Gazebo plugin
        self.publisher_ = self.create_publisher(Twist, '/cmd_vel', 10)
        timer_period = 0.1  # 10Hz
        self.timer = self.create_timer(timer_period, self.timer_callback)

    def timer_callback(self):
        msg = Twist()
        msg.linear.x = 0.2  # Move forward at 0.2 m/s
        msg.angular.z = 0.3 # Rotate at 0.3 rad/s
        self.publisher_.publish(msg)
        self.get_logger().info('Publishing: Forward and Turning')

def main(args=None):
    rclpy.init(args=args)
    node = RobotController()
    try:
        rclpy.spin(node)
    except KeyboardInterrupt:
        # Stop the robot before shutting down
        stop_msg = Twist()
        node.publisher_.publish(stop_msg)
        node.destroy_node()
        rclpy.shutdown()

if __name__ == '__main__':
    main()

File: src/my_robot_description/src/robot_controller.cpp
#include "rclcpp/rclcpp.hpp"
#include "my_robot_description/srv/set_velocity.hpp" 
#include "geometry_msgs/msg/twist.hpp" // Required for velocity commands
#include <memory>

class RobotController : public rclcpp::Node {
public:
    RobotController() : Node("robot_controller") {
        // 1. Create a Publisher to send velocity to the robot motors/simulation
        publisher_ = this->create_publisher<geometry_msgs::msg::Twist>("/cmd_vel", 10);

        // 2. Create the Service Server
        service_ = this->create_service<my_robot_description::srv::SetVelocity>(
            "set_robot_velocity",
            std::bind(&RobotController::handle_velocity_request, this, std::placeholders::_1, std::placeholders::_2));

        RCLCPP_INFO(this->get_logger(), "Robot Controller Service is ready and publishing to /cmd_vel.");
    }

private:
    void handle_velocity_request(
        const std::shared_ptr<my_robot_description::srv::SetVelocity::Request> request,
        std::shared_ptr<my_robot_description::srv::SetVelocity::Response> response) 
    {
        // Log the received values to the terminal
        RCLCPP_INFO(this->get_logger(), "Executing Command: Linear=%.2f, Angular=%.2f", 
                    request->linear_x, request->angular_z);

        // 3. Create a Twist message and fill it with the service request data
        auto move_msg = geometry_msgs::msg::Twist();
        move_msg.linear.x = request->linear_x;
        move_msg.angular.z = request->angular_z;

        // 4. Publish the message to actually move the robot
        publisher_->publish(move_msg);

        // Send confirmation back to the teleop_node
        response->success = true;
        response->message = "Velocity command published to /cmd_vel.";
    }

    // Member variables
    rclcpp::Service<my_robot_description::srv::SetVelocity>::SharedPtr service_;
    rclcpp::Publisher<geometry_msgs::msg::Twist>::SharedPtr publisher_;
};

int main(int argc, char **argv) {
    rclcpp::init(argc, argv);
    rclcpp::spin(std::make_shared<RobotController>());
    rclcpp::shutdown();
    return 0;
}